File: skills/code-optimizer/references/cpu/node/code.ts
export type RecordItem = {
  name: string;
  score: number;
};

function countVowels(text: string, pattern: RegExp): number {
  pattern.lastIndex = 0;
  return text.match(pattern)?.length ?? 0;
}

export function rankBad(records: RecordItem[]): number {
  let total = 0;
  for (const record of records) {
    const pattern = /[aeiou]/gi;
    total += countVowels(record.name, pattern) * record.score;
    const ordered = [...records].sort((left, right) =>
      left.name.localeCompare(right.name),
    );
    total += ordered[0]?.score ?? 0;
  }
  return total;
}

export function rankGood(records: RecordItem[]): number {
  const pattern = /[aeiou]/gi;
  const ordered = [...records].sort((left, right) =>
    left.name.localeCompare(right.name),
  );
  const first = ordered[0]?.score ?? 0;
  let total = 0;
  for (const record of records) {
    total += countVowels(record.name, pattern) * record.score;
    total += first;
  }
  return total;
}

File: skills/code-optimizer/references/cpu/python/code.py
import re
from dataclasses import dataclass


@dataclass(frozen=True)
class RecordItem:
  name: str
  score: int


def rank_bad(records: list[RecordItem]) -> int:
  total = 0
  for record in records:
    pattern = re.compile(r"[aeiou]", re.IGNORECASE)
    total += len(pattern.findall(record.name)) * record.score
    ordered = sorted(records, key=lambda entry: entry.name)
    total += ordered[0].score if ordered else 0
  return total


def rank_good(records: list[RecordItem]) -> int:
  pattern = re.compile(r"[aeiou]", re.IGNORECASE)
  ordered = sorted(records, key=lambda entry: entry.name)
  first = ordered[0].score if ordered else 0
  total = 0
  for record in records:
    total += len(pattern.findall(record.name)) * record.score
    total += first
  return total
